client/ayon_core/hosts/max/plugins/publish/validate_instance_in_context.py

```diff
@@ -0,0 +1,107 @@
+    so it can be disabled when needed.
+
+    Action on this validator will select invalid instances.
+    """
+
+    order = ValidateContentsOrder
+    label = "Instance in same Context"
+    optional = True
+    hosts = ["max"]
+    actions = [SelectInvalidAction, RepairAction]
+
+    def process(self, instance):
+        if not self.is_active(instance.data):
+            return
+
+        instance_node = rt.getNodeByName(instance.data.get(
+            "instance_node", ""))
+        if not instance_node:
+            return
+        asset = rt.getUserProp(instance_node, "folderPath")
+        context_asset = self.get_context_asset(instance)
+        task = rt.getUserProp(instance_node, "task")
+        context_task = self.get_context_task(instance)
+        if asset != context_asset:
+            raise PublishValidationError(
+                message=(
+                    "Instance '{}' publishes to different asset than current "
+                    "context: {}. Current context: {}".format(
+                        instance.name, asset, context_asset
+                    )
+                ),
+                description=(
+                    "## Publishing to a different asset\n"
+                    "There are publish instances present which are publishing "
+                    "into a different asset than your current context.\n\n"
+                    "Usually this is not what you want but there can be cases "
+                    "where you might want to publish into another asset or "
+                    "shot. If that's the case you can disable the validation "
+                    "on the instance to ignore it."
+                )
+            )
+        if task != context_task:
+            raise PublishValidationError(
+                message=(
+                    "Instance '{}' publishes to different task than current "
+                    "context: {}. Current context: {}".format(
+                        instance.name, task, context_task
+                    )
+                ),
+                description=(
+                    "## Publishing to a different asset\n"
+                    "There are publish instances present which are publishing "
+                    "into a different asset than your current context.\n\n"
+                    "Usually this is not what you want but there can be cases "
+                    "where you might want to publish into another asset or "
+                    "shot. If that's the case you can disable the validation "
+                    "on the instance to ignore it."
+                )
+            )
+
+    @classmethod
+    def get_invalid(cls, instance):
+        node = rt.getNodeByName(instance.data["instance_node"])
+        asset = rt.getUserProp(node, "folderPath")
+        context_asset = cls.get_context_asset(instance)
+        if asset != context_asset:
+            return instance.data["instance_node"]
```

Comment:
The get invalid logic fails to check for differing task. This already shows the issue that the validation logic is scattered - better implement the actual validation once in the plug-in so that it's consistent.

Reply:
I have rephrased a little bit